rj_gameplay/rj_gameplay/play/penalty_defense.py

```diff
@@ -1,137 +1,75 @@
 
-        # Get list of all SkillEntries from all tactics
-        skills = []
-        for tactic in self.tactics:
-            skills += tactic.tick(world_state, role_results[tactic])
+from rj_gameplay.tactic import line_tactic, goalie_tactic
 
-        # Get all role assignments
-        # SkillEntry to (list of?) RoleResult
-        skill_dict = {}
-        for tactic in self.tactics:
-            skill_dict.update(role_results[tactic])
+from typing import List, Tuple
+from enum import Enum, auto
 
-        return (skill_dict, skills)
 
-    def is_done(self, world_state):
-        # returns done when all tactics are done
-        # TODO: change all is_done() to use all()?
-        return all([tactic.is_done(world_state) for tactic in self.tactics])
+class State(Enum):
+    INIT = auto()
+    ACTIVE = auto()
+    DONE = auto()
 
 
-class PenaltyDefense(play.IPlay):
-    """For when we don't have the ball and are trying to stop the opponent from scoring."""
+class PenaltyDefense(stp.play.Play):
+    """
+    Play consistas of:
+                    -5 wallers
+                    -1 goalie
+    The goalie will defend the opponents penalty kick.
+    """
 
     def __init__(self):
-        self.tactics = [
-            goalie_tactic.GoalieTactic(),
-            move_tactic.Move((1.6, 9.0)),
-            move_tactic.Move((1.9, 9.0)),
-            move_tactic.Move((2.2, 9.0)),
-            move_tactic.Move((2.5, 9.0)),
-            move_tactic.Move((2.8, 9.0)),
-        ]
-
-        self.role_assigner = NaiveRoleAssignment()
-
-    def compute_props(self, prev_props):
-        pass
+        super().__init__()
+        self._state = State.INIT
 
     def tick(
         self,
-        world_state: rc.WorldState,
-        prev_results: role.assignment.FlatRoleResults,
-        props,
-    ) -> Tuple[
-        Dict[Type[tactic.SkillEntry], List[role.RoleRequest]],
-        List[tactic.SkillEntry],
-    ]:
-
-        # Get role requests from all tactics and put them into a dictionary
-        role_requests: play.RoleRequests = {
-            tactic: tactic.get_requests(world_state, None) for tactic in self.tactics
-        }
-
-        # Flatten requests and use role assigner on them
-        flat_requests = play.flatten_requests(role_requests)
-        flat_results = self.role_assigner.assign_roles(
-            flat_requests, world_state, prev_results
-        )
-        role_results = play.unflatten_results(flat_results)
+        world_state: stp.rc.WorldState,
+    ) -> List[RobotIntent]:
+        if self._state == State.INIT:
+            self.prioritized_tactics.append(goalie_tactic.GoalieTactic(world_state, 0))
+            self.prioritized_tactics.append(line_tactic.LineTactic(world_state, 5))
+            self.assign_roles(world_state)
+            self._state = State.ACTIVE
+            return self.get_robot_intents(world_state)
+        elif self._state == State.ACTIVE:
+            if self.prioritized_tactics[-1].is_done(world_state):
```

Comment:
Why are you checking if one of the wallers is done? Shouldn't it be based on if the goalie has the ball or if a goal has been scored or something.

Reply:
^ goalie has ball would be easiest to implement. To be honest, this play doesn't really need an end since the ref will signal a penalty end and the play selector will pick a different play regardless.